import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/backupuploader/GcpBackupUploader.java

```diff
@@ -1,25 +1,129 @@
+  }
+
+  public GcpBackupUploader(String uploadFilePath, Storage storage) throws IOException {
+    super(uploadFilePath);
+    parseUploadFilePath(getUploadFilePath());
+    this.storage = storage;
   }
 
   @Override
-  public void doWriteBackup(InputStream inputStream, String uniqueFileName) throws BackupException {
-    
+  public void doWriteBackup(InputStream backupData, String uniqueFileName)
+          throws BackupException {
+    try {
+      if (backupData == null) {
+        throw new BackupException("Backup data is null");
+      }
+      validatePathParameter(uniqueFileName, "unique file name");
+      BlobId blobId = BlobId.of(bucketName, getFullUploadObject(uniqueFileName));
+      BlobInfo blobInfo = BlobInfo.newBuilder(blobId).build();
+      storage.createFrom(blobInfo, backupData);
+    } catch (Exception e) {
+      throw new BackupException("Error with writing backup file: " + e.getMessage(), e);
+    }
   }
 
   @Override
   public InputStream doReadBackup(String uniqueFileName) throws BackupException {
-    return null;
+    try {
+      validatePathParameter(uniqueFileName, "unique file name");
+      Blob blob = storage.get(BlobId.of(bucketName, getFullUploadObject(uniqueFileName)));
+      ReadChannel channel = blob.reader();
+      return Channels.newInputStream(channel);
+    } catch (Exception e) {
+      throw new BackupException("Error with reading backup file: " + e.getMessage(), e);
+    }
   }
 
   @Override
   public void doRemoveBackup(String uniqueFileName) throws BackupException {
+    try {
+      validatePathParameter(uniqueFileName, "unique file name");
+      storage.delete(bucketName, getFullUploadObject(uniqueFileName));
+    } catch (Exception e) {
+      throw new BackupException("Error with removing backup file: " + e.getMessage(), e);
+    }
+  }
+
+  private void parseUploadFilePath(String uploadFilePath) throws GcpUriParseException {
+    try {
+      validatePathParameter(uploadFilePath, "upload file path");
+      List<String> segments = new URIBuilder().setPath(getUploadFilePath()).getPathSegments();
+      projectName = segments.get(2);
+      bucketName = segments.get(3);
+      uploadObject = new URIBuilder()
+              .setPathSegments(segments.subList(4, segments.size()))
+              .getPath().substring(1);
+      validatePathParameter(projectName, "project name");
+      validatePathParameter(bucketName, "bucket name");
+      validatePathParameter(uploadObject, "upload object");
+    } catch (Exception e) {
+      throw new GcpUriParseException("Invalid upload path: " + e.getMessage(), e);
+    }
+  }
+
+  public String getProjectName() {
+    return projectName;
+  }
+
+  public String getBucketName() {
+    return bucketName;
+  }
+
+  public String getUploadObject() {
+    return uploadObject;
+  }
+
+  public Credentials getCredential(String env) throws IOException {
+    return GoogleCredentials
+            .fromStream(new FileInputStream(System.getenv(env)));
+  }
```

Comment:
We get credentials in the export adapter here: https://github.com/GoogleCloudPlatform/healthcare-dicom-dicomweb-adapter/blob/e72628beaf645b70264f377d782a8b6315c8f00c/export/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/ExportAdapter.java#L56
Could we use the same method here? to avoid relying on an env variable